packages/node/src/utils/promise-queue.ts

```diff
@@ -0,0 +1,56 @@
+interface Deferred<T> {
+  promise: Promise<T>;
+  resolve: (value: T | PromiseLike<T>) => void;
+  reject: (err?: unknown) => void;
+}
+
+interface Node<T> {
+  value: T;
+  tail: Promise<Node<T>>;
+}
+
+const createDeferred = <T>(): Deferred<T> => {
+  const deferred = {
+    promise: Promise.resolve<T>(null as any),
+    resolve: (_value: T) => {},
+    reject: (_reason?: unknown) => {},
+  };
+
+  deferred.promise = new Promise<T>((resolve, reject) => {
+    deferred.resolve = resolve;
+    deferred.reject = reject;
+  });
+
+  return deferred;
+};
+
+export class PromiseQueue<T> {
```

Comment:
This data structure is necessary to take the push-based child process into a pull-based async iterator